### ISSUE TITLE
Generating migrations recreates materialized views

### ISSUE THREAD

**Issue by feimosi** [typeorm/typeorm#6626]
**Issue type:**

[X] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[X] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[ ] `latest`
[ ] `@next`
[X] `0.2.25` (latest at the time of reporting this)

**Steps to reproduce or a small repository showing the problem:**

1. Create a materialized view
```ts
@ViewEntity({
  expression:
  `SELECT ...`,
  materialized: true,
})
export class UserTotal {
  @ViewColumn()
  @IsDefined()
  total!: number;

  ...
}
```

2. Run `typeorm migration:generate`

3. Run `typeorm migration:run`

4. Run `typeorm migration:generate` again

TypeORM doesn't see materialized views and always regenerates them.

After some digging I found what's wrong.
The following method `loadViews`:
https://github.com/typeorm/typeorm/blob/327144a31479d4a2de2222309b216959249039d7/src/driver/postgres/PostgresQueryRunner.ts#L1312
looks for database views inside `information_schema`, **however** PostgreSQL stores materialized views in a different way, namely inside [`pg_matviews`](https://www.postgresql.org/docs/current/view-pg-matviews.html).

In the meantime, for anyone struggling with this issue, I prepared a workaround overriding that method:
```ts
import { PostgresQueryRunner } from 'typeorm/driver/postgres/PostgresQueryRunner';
import { View } from 'typeorm/schema-builder/view/View';

type LoadViewsParams = Parameters<typeof PostgresQueryRunner.prototype['loadViews']>;

const originalLoadViews = PostgresQueryRunner.prototype['loadViews'];

PostgresQueryRunner.prototype['loadViews'] = async function loadViews(...args: LoadViewsParams) {
  // Based on:
  // https://github.com/typeorm/typeorm/blob/327144a31479d4a2de2222309b216959249039d7/src/driver/postgres/PostgresQueryRunner.ts#L1329
  const query = `SELECT "t".* FROM ${this['escapePath'](this['getTypeormMetadataTableName']())} "t" ` +
  'INNER JOIN "pg_matviews" "v" ON "v"."schemaname" = "t"."schema" ' +
  'AND "v"."matviewname" = "t"."name" WHERE "t"."type" = \'VIEW\'';

  const queryResult = await this.query(query) as object[];

  const materializedViews = queryResult.map((databaseView: any) => {
    const view = new View();
    const schema = '';
    view.name = this.driver.buildTableName(databaseView['name'], schema);
    view.expression = databaseView['value'];

    return view;
  });

  const views = await originalLoadViews.apply(this, args);

  return materializedViews.concat(views);
};
```

<!--
    To answer those questions you need to put "x" inside the square brackets, for example:
    [x] `mysql`
    [ ] `postgres`

    Also, please format your code properly (by taking code blocks into ```ts .... ```)
--!>

**Comment by nimmortal** [comment]
Hello guys,

Any update regarding materialized view migration?
Very annoying bug.

**Comment by pleportz** [comment]
Hello

I encountered the same issue. As a temporary workaround, I followed this process:

1. Define the materialized view class
```
@ViewEntity({
  name: '<my_table_name>',
  materialized: true,
  synchronize: true,
  expression: `<MY SQL QUERY>`,
})
export class MyClassName {...}
```
2. Run `typeorm migration:generate`
2. Run `typeorm migration:run`
3. Switch to `synchronize: false,` in `@ViewEntity`
4. Re-run `typeorm migration:generate` : at this point, no migration is generated

Would be great to have a more robust solution integrated into typeorm, as described by @feimosi
